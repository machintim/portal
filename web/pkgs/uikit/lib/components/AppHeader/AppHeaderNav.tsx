import { memo, HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';

export const NAME = 'ucl-timetable-app-header__nav';

export interface AppHeaderNavProps
  extends HTMLAttributes<HTMLDivElement> {
  testId?: string;
}

const AppHeaderNav = ({
  testId = NAME,
  className,
  children,
  ...props
}: AppHeaderNavProps) => {
  const baseStyle = css`
    flex: 1;
    height: 100%;
    display: flex;
    align-items: center;
  `;

  const style = cx(NAME, baseStyle, className);

  return (
    <div
      className={style}
      data-testid={testId}
      {...props}
    >
      {children}

      {/* <img src={burgerIcon} className={menuIconStyle}/> */}
    </div>
  );
};

export default memo(AppHeaderNav);
