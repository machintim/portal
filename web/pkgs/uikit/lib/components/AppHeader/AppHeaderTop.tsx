import { memo, HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import { UclLogo, useTheme } from '../../';

export const NAME = 'ucl-timetable-app-header__top';

export const HEIGHT = 56;

export interface AppHeaderTopProps
  extends HTMLAttributes<HTMLDivElement> {
  uclLogo?: boolean;
  logoClassName?: string;
  testId?: string;
}

const AppHeaderTop = ({
  uclLogo = true,
  logoClassName,
  testId = NAME,
  className,
  children,
  ...props
}: AppHeaderTopProps) => {
  const [theme] = useTheme();

  const baseStyle = css`
    height: ${HEIGHT}px;
    padding: 0 ${theme.padding.p16};
    display: flex;
  `;

  const style = cx(NAME, baseStyle, className);

  const uclLogoBaseStyle = css`
    display: block;
    margin-left: auto;
    height: 100%;
    color: ${theme.color.display
      .greenDark}; // default header/logo colour
  `;

  const uclLogoStyle = cx(uclLogoBaseStyle, logoClassName);

  return (
    <div
      className={style}
      data-testid={testId}
      {...props}
    >
      {children}

      {uclLogo && <UclLogo className={uclLogoStyle} />}
    </div>
  );
};

export default memo(AppHeaderTop);
