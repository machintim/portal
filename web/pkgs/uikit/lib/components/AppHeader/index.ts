export { default } from './AppHeader';
export { HEIGHT as APP_HEADER_HEIGHT } from './AppHeader';
export type { AppHeaderProps } from './AppHeader';
