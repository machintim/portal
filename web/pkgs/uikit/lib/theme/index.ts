export { default as theme } from './defaultTheme';
export type {
  ThemeType,
  DisplayColoursType,
  BlackAndWhiteColoursType,
} from './defaultTheme';
export {
  default as useTheme,
  ThemeContextProvider,
} from './useTheme';
