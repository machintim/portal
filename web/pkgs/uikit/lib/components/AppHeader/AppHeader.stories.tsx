import type { Meta, StoryObj } from '@storybook/react';
import { css } from '@emotion/css';

import AppHeader from './AppHeader';
import Breadcrumbs from '../Breadcrumbs/Breadcrumbs';
import AppMenu from '../AppMenu/AppMenu';
import Icon from '../Icon/Icon';
import { useTheme } from '../../theme';

const meta = {
  title: 'Components/Deprecated/AppHeader',
  component: AppHeader,
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof AppHeader>;

export default meta;

type Story = StoryObj<typeof meta>;

export const WithBreadcrumbs: Story = {
  name: 'With breadcrumbs',

  render: () => {
    const [theme] = useTheme();

    // This ought to be done in`<Breadcrumb>`or in `<AppHeader.Bottom>`!
    const invertedTextStyle = css`
      color: ${theme.color.text.inverted};
    `;

    return (
      <AppHeader className={invertedTextStyle}>
        <AppHeader.Top />
        <AppHeader.Bottom>
          <Breadcrumbs>
            <Breadcrumbs.Breadcrumb uri='/'>Home</Breadcrumbs.Breadcrumb>
          </Breadcrumbs>
        </AppHeader.Bottom>
      </AppHeader>
    );
  },
};

export const WithoutBreadcrumbs: Story = {
  name: 'Without breadcrumbs',
  render: () => {
    return (
      <AppHeader>
        <AppHeader.Top />
        <AppHeader.Bottom />
      </AppHeader>
    );
  },
};

export const WithMultipleBreadcrumbs: Story = {
  name: 'With multiple breadcrumbs',
  render: () => {
    const [theme] = useTheme();
    const invertedTextStyle = css`
      color: ${theme.color.text.inverted};
    `;
    return (
      <AppHeader>
        <AppHeader.Top />
        <AppHeader.Bottom className={invertedTextStyle}>
          <Breadcrumbs>
            <Breadcrumbs.Breadcrumb uri='/'>Home</Breadcrumbs.Breadcrumb>
            <Breadcrumbs.Breadcrumb uri='/showcase'>
              Component Library
            </Breadcrumbs.Breadcrumb>
            <Breadcrumbs.Breadcrumb uri='/showcase/app-header'>
              App Header
            </Breadcrumbs.Breadcrumb>
          </Breadcrumbs>
        </AppHeader.Bottom>
      </AppHeader>
    );
  },
};

export const WithAppMenu: Story = {
  name: 'With AppMenu',
  render: () => {
    const [theme] = useTheme();

    const centerVerticalStyle = css`
      display: flex;
      align-items: center;
    `;

    const invertedTextStyle = css`
      color: ${theme.color.text.inverted};
    `;

    return (
      <AppHeader>
        <AppHeader.Top className={centerVerticalStyle}>
          <AppMenu>
            <AppMenu.Trigger>
              <Icon.Menu />
            </AppMenu.Trigger>
            <AppMenu.Content>
              <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
                Item 1
              </AppMenu.MenuItem>
              <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
                Item 2
              </AppMenu.MenuItem>
              <AppMenu.MenuItem icon={<Icon.Calendar size={20} />}>
                Item 3
              </AppMenu.MenuItem>
              <AppMenu.Divider />
              <AppMenu.MenuItem icon={<Icon.Settings size={20} />}>
                Settings
              </AppMenu.MenuItem>
              <AppMenu.MenuItem icon={<Icon.HelpCircle size={20} />}>
                Help
              </AppMenu.MenuItem>
              <AppMenu.Divider />
              <AppMenu.MenuItem icon={<Icon.ArrowLeft size={20} />}>
                Log out
              </AppMenu.MenuItem>
            </AppMenu.Content>
          </AppMenu>
        </AppHeader.Top>
        <AppHeader.Bottom className={invertedTextStyle}>
          <Breadcrumbs.Breadcrumb uri='/'>Home</Breadcrumbs.Breadcrumb>
        </AppHeader.Bottom>
      </AppHeader>
    );
  },
};
