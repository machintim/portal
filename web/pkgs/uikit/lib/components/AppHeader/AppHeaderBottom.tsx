import { memo, HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import { useTheme } from '../../../lib';

export const NAME = 'ucl-timetable-app-header__bottom';

export const HEIGHT = 48;

export interface AppHeaderBottomProps
  extends HTMLAttributes<HTMLDivElement> {
  testId?: string;
}

const AppHeaderBottom = ({
  testId = NAME,
  className,
  children,
  ...props
}: AppHeaderBottomProps) => {
  const [theme] = useTheme();

  const baseStyle = css`
    height: ${HEIGHT}px;
    padding: 0 ${theme.padding.p16};
    line-height: 48px;
    display: flex;
    align-items: center;
    background-color: ${theme.color.display
      .greenDark}; // default header/logo colour
    color: inherit;
  `;

  const style = cx(NAME, baseStyle, className);

  return (
    <div
      className={style}
      data-testid={testId}
      {...props}
    >
      {children}
    </div>
  );
};

export default memo(AppHeaderBottom);
