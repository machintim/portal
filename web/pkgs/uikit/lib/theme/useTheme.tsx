// We will probably move away from context/provider to store/hook for accessing theme.

import {
  createContext,
  useState,
  useContext,
  ReactNode,
} from 'react';

import theme, { ThemeType } from './defaultTheme';

const useThemeState = () => useState(theme as ThemeType);

const ThemeContext = createContext<ReturnType<
  typeof useThemeState
> | null>(null);

export const ThemeContextProvider: React.FC<{
  children: ReactNode;
}> = ({ children }) => (
  <ThemeContext.Provider value={useThemeState()}>
    {children}
  </ThemeContext.Provider>
);

const useTheme = () => {
  const value = useContext(ThemeContext);
  if (value === null) {
    throw new Error('Please add ThemeContextProvider');
  }
  return value;
};

export default useTheme;
