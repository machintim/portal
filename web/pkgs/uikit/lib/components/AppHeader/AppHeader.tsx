import { memo, HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import { useTheme } from '../../../lib';
import AppHeaderTop, {
  HEIGHT as APP_HEADER_TOP_HEIGHT,
} from './AppHeaderTop';
import AppHeaderBottom, {
  HEIGHT as APP_HEADER_BOTTOM_HEIGHT,
} from './AppHeaderBottom';
import AppHeaderNav from './AppHeaderNav';

export const NAME = 'ucl-timetable-app-header';

export const Z_INDEX = 3;

export const HEIGHT =
  APP_HEADER_TOP_HEIGHT + APP_HEADER_BOTTOM_HEIGHT;

export interface AppHeaderProps
  extends HTMLAttributes<HTMLDivElement> {
  fixed?: boolean;
  sticky?: boolean;
  testId?: string;
}

const AppHeader = ({
  fixed = true,
  sticky = false,
  testId = NAME,
  className,
  children,
  ...props
}: AppHeaderProps) => {
  const [theme] = useTheme();

  const baseStyle = css`
    background-color: ${theme.color.neutral.white};
    color: ${theme.color.text.primary};
    font-family: ${theme.font.family.primary};
  `;

  const fixedStyle = css`
    position: fixed;
    left: 0;
    top: 0;
    width: 100%;
    z-index: ${Z_INDEX};
  `;

  const stickyStyle = css`
    position: sticky;
    top: -${APP_HEADER_TOP_HEIGHT}px;
  `;

  const style = cx(
    NAME,
    baseStyle,
    fixed && fixedStyle,
    sticky && stickyStyle,
    className
  );

  return (
    <div
      className={style}
      data-testid={testId}
      {...props}
    >
      {children}
    </div>
  );
};

const MemoAppHeader = memo(AppHeader);

export interface IHeaderSubComponents {
  Top: typeof AppHeaderTop;
  Bottom: typeof AppHeaderBottom;
  Nav: typeof AppHeaderNav;
}

const AppHeaderWithSubComponents =
  MemoAppHeader as typeof MemoAppHeader &
    IHeaderSubComponents;

AppHeaderWithSubComponents.Top = AppHeaderTop;
AppHeaderWithSubComponents.Bottom = AppHeaderBottom;
AppHeaderWithSubComponents.Nav = AppHeaderNav;

export default AppHeaderWithSubComponents;
