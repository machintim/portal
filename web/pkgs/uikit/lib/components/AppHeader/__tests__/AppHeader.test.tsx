import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import AppHeader from '../AppHeader';
import { ThemeContextProvider } from '../../../theme/useTheme';

describe('AppHeader', () => {
  // Snapshot tests

  test('snapshot: no props', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <AppHeader />
      </ThemeContextProvider>
    );
    expect(
      renderResult.container.firstChild
    ).toMatchSnapshot();
  });
});
