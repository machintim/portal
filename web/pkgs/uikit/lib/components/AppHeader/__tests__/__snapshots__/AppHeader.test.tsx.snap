// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppHeader > snapshot: no props 1`] = `
<div
  class="ucl-timetable-app-header css-14i4gnx"
  data-testid="ucl-timetable-app-header"
/>
`;
